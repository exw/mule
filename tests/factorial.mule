let factorial =
  fn n. match n with
  | 0 -> 1
  | n -> mul n (factorial (sub n 1))
  end
in
factorial 5
