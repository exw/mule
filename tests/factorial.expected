
Desugared:

	let factorial = fix/let fn factorial. fn n. 
	                                              match-lam
	                                              | 0 -> 1
	                                              | fn n. mul n (factorial (sub n 1))
	                                              end n in
	factorial 5

inferred type:

	int

Runtime term:

	(fn. v0 5) (fix/let fn. fn. 
	                              match _ with
	                              | 0 -> 1
	                              | 
	                                  fn. <built-in function> v0 (v2 (<built-in function> v0 1))
	                              end v0)

Evaluated:

	120
